File: python/shapes/cylinder_adapter.py
import math
from typing import Tuple

from .ishape_adapter import Cell, IShapeAdapter


class CylinderAdapter(IShapeAdapter):
    """Adapter mapping grid cells onto a cylinder with top and bottom faces."""

    def __init__(self, size: int = 10, radius: float | None = None) -> None:
        self.size = size
        self.radius = radius if radius is not None else size / (2 * math.pi)

    def get_grid_size(self) -> Tuple[int, int]:
        return self.size, self.size

    def get_face_count(self) -> int:
        return 3

    def to_world(self, cell: Cell) -> Tuple[float, float, float]:
        angle = (cell.u + 0.5) * (2 * math.pi / self.size)
        offset = self.size / 2 - 0.5
        if cell.face == 0:  # side
            x = self.radius * math.cos(angle)
            z = self.radius * math.sin(angle)
            y = offset - cell.v
            return (x, y, z)
        if cell.face == 1:  # top
            r = (cell.v + 0.5) * (self.radius / self.size)
            x = r * math.cos(angle)
            z = r * math.sin(angle)
            y = offset
            return (x, y, z)
        if cell.face == 2:  # bottom
            r = (cell.v + 0.5) * (self.radius / self.size)
            x = r * math.cos(angle)
            z = r * math.sin(angle)
            y = -offset
            return (x, y, z)
        return (0.0, 0.0, 0.0)

    def wrap(self, cell: Cell, direction: str) -> Cell:
        face = cell.face
        u = cell.u
        v = cell.v
        last = self.size - 1
        if face == 0:
            if u < 0:
                u = last
            if u > last:
                u = 0
            if v < 0 and direction == "up":
                return Cell(1, u, 0)
            if v > last and direction == "down":
                return Cell(2, u, 0)
        elif face == 1:
            if u < 0:
                u = last
            if u > last:
                u = 0
            if v < 0 and direction == "down":
                return Cell(0, u, 0)
        elif face == 2:
            if u < 0:
                u = last
            if u > last:
                u = 0
            if v < 0 and direction == "up":
                return Cell(0, u, last)
        return Cell((face), (u + self.size) % self.size, (v + self.size) % self.size)

File: src/shapes/CylinderAdapter.ts
import * as THREE from 'three';
import type { Cell, Direction } from '../core/Grid';
import type { IShapeAdapter } from './IShapeAdapter';

/**
 * Projects grid coordinates onto a cylinder with top and bottom caps.
 * Face 0 is the curved side, face 1 the top, and face 2 the bottom.
 */
export class CylinderAdapter implements IShapeAdapter {
  private radius: number;

  constructor(private size = 10, radius?: number) {
    this.radius = radius ?? size / (2 * Math.PI);
  }

  getGridSize() {
    return { u: this.size, v: this.size };
  }

  getFaceCount() {
    return 3;
  }

  /**
   * Convert a grid cell to world space on the cylinder surface.
   *
   * `u` rotates around the cylinder's axis while `v` moves along the height.
   * For the caps, `v` represents radial distance from the center.
   */
  toWorld(cell: Cell): THREE.Vector3 {
    const angle = (cell.u + 0.5) * (2 * Math.PI / this.size);
    const offset = this.size / 2 - 0.5;
    switch (cell.face) {
      case 0: // side
        return new THREE.Vector3(
          this.radius * Math.cos(angle),
          offset - cell.v,
          this.radius * Math.sin(angle)
        );
      case 1: // top
        {
          const r = (cell.v + 0.5) * (this.radius / this.size);
          return new THREE.Vector3(
            r * Math.cos(angle),
            offset,
            r * Math.sin(angle)
          );
        }
      case 2: // bottom
        {
          const r = (cell.v + 0.5) * (this.radius / this.size);
          return new THREE.Vector3(
            r * Math.cos(angle),
            -offset,
            r * Math.sin(angle)
          );
        }
      default:
        return new THREE.Vector3();
    }
  }

  /**
   * Wrap coordinates around the cylinder and between faces.
   *
   * When leaving the side at the top or bottom the cell transitions to the
   * respective cap. Horizontal movement around the side wraps modulo the grid
   * width.
   */
  wrap(cell: Cell, dir: Direction): Cell {
    const { face } = cell;
    let { u, v } = cell;
    const last = this.size - 1;
    switch (face) {
      case 0: // side
        if (u < 0) u = last;
        if (u > last) u = 0;
        if (v < 0 && dir === 'up') return { face: 1, u, v: 0 };
        if (v > last && dir === 'down') return { face: 2, u, v: 0 };
        break;
      case 1: // top
        if (u < 0) u = last;
        if (u > last) u = 0;
        if (v < 0 && dir === 'down') return { face: 0, u, v: 0 };
        if (v > last && dir === 'up') return { face: 0, u, v: last };
        break;
      case 2: // bottom
        if (u < 0) u = last;
        if (u > last) u = 0;
        if (v < 0 && dir === 'up') return { face: 0, u, v: last };
        if (v > last && dir === 'down') return { face: 0, u, v: 0 };
        break;
    }
    u = (u + this.size) % this.size;
    v = (v + this.size) % this.size;
    return { face, u, v };
  }
}
